pinot/generative/torch_gvae/model.py

```diff
@@ -0,0 +1,84 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from pinot.generative.torch_gvae.layers import GraphConvolution
+
+
+class GCNModelVAE(nn.Module):
+    """Graph convolutional neural networks for VAE
+    """
+    def __init__(self, input_feat_dim, hidden_dim1, hidden_dim2, dropout):
+        """ Construct a VAE with GCN
+        """
+        super(GCNModelVAE, self).__init__()
+        self.gc1 = GraphConvolution(input_feat_dim, hidden_dim1, dropout, act=F.relu)
+        self.dc = InnerProductDecoder(dropout, act=lambda x: x) # Decoder
+        self.output_regression = torch.nn.ModuleList([
+            GraphConvolution(hidden_dim1, hidden_dim2, dropout, act=lambda x: x),
+            GraphConvolution(hidden_dim1, hidden_dim2, dropout, act=lambda x: x),
+        ])
+
+    def forward(self, x, adj):
```

Comment:
if we were to depend on DGL anyway we might not need to see adjacency matrix too often in the scripts expect when we output it from the generative model.

Reply:
@yuanqing-wang I'm not too sure what you mean here. A forward pass in GVAE requires the node feature vectors `x` and the adjacency matrix `adj`. We can certainly combine them into one argument `g = { 'x' : x, 'adj' : adj}` (is that what you were trying to point out, that `adj` doesn't have to be explicitly defined as a separate argument to pass into `forward`?)

Reply:
Ah I'm not nitpicking on the way we feed arguments into functions. My point is that `adj` is required here only because the GitHub repo you ported chose to express GraphConv in this way. This however could be expressed a lot simpler if we used `dgl`.

Reply:
Also it would require some minor headache if you wanted to batch and unbatch adjacency matrix. (you'll need to either slice the sparse matrix or do some sort of slice along diagnal) Nothing too complicated, but DGL does that for you.

Reply:
Guys, I will restate this:
I strongly support also trying non-dgl representations for a standalone effort here.
Yianqing, what Duc is doing is exploring different models and codebases to write them up.
I explicitly want to let him explore what is out there in terms of both models and code that runs to explore these APIs. I think what you are suggesting is quite a likely outcome of this process, but I want it to be the outcome, not an axiom.

Dgl, like any library, may have edgecases where it cannot express stuff that we may find we need. Let Duc play with this stuff for a while with a standalone module that just tests different models and codebases.

Reply:
Also there is ```pytorch geometric``` and a variety of other ways to express graphs. We should not prematurely restrict our explorations to dgl, though I like dgl and was the person who originally suggested it as a quick way to get into pytorch graph stuff, if you remember.

Reply:
@yuanqing-wang I see what you mean now. I will certainly try to see if DGL can be incorporated here. That would probably make the code more robust since DGL has been tested well.

Reply:
@dnguyen1196 before you refactor out of this:
Can you please leave empirical information about how well this handwritten code works (empirically, i.e. numbers) vis a vis a library like dgl?
You would be surprised how often I have seen things diverge between low level pytorch code and output of libraries.

Reply:
@dnguyen1196 @karalets 

How about this:

for this model and the future generative models. we don't explicitly depend on dgl or anything. rather, we leave the representation (**g -> h **) bit empty and only implement methods like `generate` and `inference` methods. If there's anything unique to the **g -> h ** part of some model, leave that function out of the main object so that we can plug in our favorite representations.
- as of now, the only parts that invokes `dgl` libraries are the representation part. in general, this should be a neurally parametrized, trainable function that takes a graph as input and a `(n_nodes, hidden_dimension)` representation as output. so long as a function satisfies this, this could be plugged in as representation. (of course now the graph itself is also a DGL object but we can implement alternatives later.)

My main point is that, generative models typically refer to, and are characterized by, the function that translates a latent code to a graph. so it should be structured as such.